Auto-conversion

cailmdaley opened the pull request.

cailmdaley committed:
jl -> py rework + autoconvert

cailmdaley commented:
Half-finished attempt at setting up the auto-convert machinery. Julia -> Python is working pretty well, see `example.jl` to see how things are called.

cailmdaley committed:
turned interface into Julia Pkg, auto-convert works

cailmdaley commented:
Auto-conversion now works in both directions, hopefully just minor details to clean up at this point!

cailmdaley commented on CMBLensingSPT3GInterface/src/CMBLensingSPT3GInterface.jl:
```diff
@@ -0,0 +1,223 @@
+    pytype_mapping(py"dict", FieldTuple)
+end
+
+Base.getindex(f::FlatField, s::String) = f[s == "T" ? :I : Symbol(s)]
```
This is very convenient, but maybe a bit invasive...

marius311 replied:
I think this is OK for now, it at least leaves a clear rule, its `:I,:E,:B`  on the Julia side and `"T", "E", "B"` on the Python side.

Fwiw my choice on the Julia side is because not all S0 fields are temperature, so it felt really wierd writing things like `ϕ.Tx`, so switched to slightly more generic `I` for intensity, like the intensity of the grav lensing potential or the intensity oft the deflection field somehow feels less obvoiusly wrong that the temperature of the grav lensing potential / deflection.

cailmdaley replied:
Funny, I always thought it was to avoid confusion with Julia's `f(x::T) where T` convention

cailmdaley commented on CMBLensing_spt3g_interface/CMBLensing_spt3g_interface.py:
```diff
@@ -1,156 +0,0 @@
-### initialize PyJulia
-import sys
-import julia
-from .julia_jlcommand import jl
```
Now that everything is in a single .jl file, what do we do about the `jl` command? The simplest thing would be to stick its contents in `CMBLensingSPT3GInterface.__init__()` maybe?

marius311 replied:
I'd keep it its own separate file, but can certainly import it from that `__init__`.

cailmdaley replied:
Aren't there path issues with importing from `__init__`? I was under the impression that `py"import .julia_jlcommand"` in the `__init__` of a Julia package wouldn't know to look in the package's `src` directory.

marius311 replied:
I don't think I understood your question right the first time, but anyway I think the Python stuff is set up ok now. After following the install instructions in the readme I added, its just

```python
>>> from CMBLensingSPT3GInterface import jl

>>> jl("FlatMap(rand(2,2))") # will come out a FlatSkyMap in Python, etc...
```

cailmdaley commented on example.jl:
```diff
@@ -0,0 +1,40 @@
+using Pkg; Pkg.activate("CMBLensingSPT3GInterface")
```
Unfortunately, adding the PyCall `pytype_mapping` PR as a package dependency isn't enough to get auto-convert to work, I think since the PyCall pinned in the user's environment is the one that actually sets the auto-convert precedence in the wild. This is only a minor inconvenience, but it is a bit unsatisfying..

marius311 replied:
Yea, there's no way for a package to specify a dependency on a non-tagged version. We can either 1) from the repo which _uses_ this package, pin the right PyCall version in that repo's Manifest or 2) include the code from my PyCall PR  in this package and just overrwrite the necessary function. I'd just go with 1 for now since pretty much we only ever use this from ultradeep100d or delensing, where we'd be activating that project's Manifest. But we could do 2 if somehow this get wider spread use and that PR doesnt get merged.

cailmdaley replied:
Ah, so the problem is that the dependency is un-tagged, good to know. I would vote for 2), since I use this interface independently from any specific project pretty frequently. It's just copying over [these lines](https://github.com/JuliaPy/PyCall.jl/blob/b66bda1d10ac80dddd22c4d2abf47385269724f5/src/conversions.jl#L800-L821), right?

cailmdaley replied:
But maybe there are other drawbacks to 2)?

marius311 replied:
Only draw back will be a function redefinition warning, but I'll switch to going with option 2.

marius311 replied:
(I was able to make it so that if you _do_ have my PR checked out, you wont get the function redefinition warning)

cailmdaley replied:
Great!

cailmdaley commented on CMBLensingSPT3GInterface/src/CMBLensingSPT3GInterface.jl:
```diff
@@ -0,0 +1,223 @@
+module CMBLensingSPT3GInterface
```
I wonder if there is a better name-- the `SPT3GI` part is a bit intense..

marius311 replied:
CMBLensingSPT3GInterface was exactly what I would have named it, so I suppose thats a sign its the right thing.

marius311 committed:
tweak a few conversions and add some tests

marius311 commented:
Finally took a detailed look. For the most part its great, thanks, I just added a few tweaks. Mainly, 

* I think the FFT flip logic before was only correct for FFT's which were all real, now it works for anything. 
* It seemed like in some places conversion factors were getting applied in one direction but not the other, such that a field wasn't surviving a round trip to Python, not sure if that was intentional. I made it consistent and added tests to reflect that too. 
* As a stupid hack, note the `0"""` that fixes the syntax highlighting thing, I cant think of anything better. 
* I changed the name `Parent` which was always a big cryptic to me to `similar_FlatSkyMap` which mirrors Julia `similar`, in that it creates and new and empty object similar to the thing you pass in. 

I think the main question I have left is, there are 3 objects which have and equivlanet on the 3G side, FlatSkyMap, MapSpectrum2D, and MapSpectrumTEB. Currently we only autoconvert the first two, should we also do the latter? I would say yes, let me know what you think.

marius311 commented:
I guess the main thing left is maybe units. I like your suggestion above, but maybe we should be more generic and call it `py"unitless(f)"` and `jl("unitless(f)")'`, where by default (as we are I think) we assume a unitful FFT and convert between the two conventions. 

However, this makes me think, 1) a transfer function will have no units, 2) the FFT of a field will have some units, and 3) a covariance like `diag(Cf)` should have those units squared, so it seems like we should technically need 3 different cases? I'm slightly confused why it doesn't appear we've needed that thus far. Maybe we only ever used 1 and 3, and otherwise were doing conversion in map space.

cailmdaley commented on CMBLensingSPT3GInterface/src/CMBLensingSPT3GInterface.jl:
```diff
@@ -208,15 +223,15 @@ end
 ### py -> jl
 ############
 function Base.convert(::Type{FlatIEBFourier}, f::PyObject)
-    @assert pytypeof(f) == py"MapSpectraTEB "
```
I've found that without this trailing space, Atom's syntax highlighting incorrectly reads `py"...B"` as a an unclosed `b"...` `Base.CodeUnits` string and ruins highlighting for the rest of the document. I may add these spaces back in unless you know of a workaround?

marius311 replied:
Which versions of Atom stuff are you on? Its working ok for me: 

![image](https://user-images.githubusercontent.com/1510968/88489605-f9478e80-cf49-11ea-97e2-700d371c3a04.png)

cailmdaley commented on CMBLensingSPT3GInterface/src/CMBLensingSPT3GInterface.jl:
```diff
@@ -208,15 +223,15 @@ end
-function MapSpectraTEB(f)
-    frame = Frame(Fourier(f), "TEB")
+function MapSpectraTEB(f::FlatIEBFourier)
+    frame = Frame(f, "TEB")
```
Is there a reason to restrict a user from going from _any_ `FlatS02` to a python `MapSpectraTEB` in one step? Calling e.g. `MapSpectraTEB(f::FlatIQUMap)`(which worked before) is convenient

cailmdaley commented:
Thanks for these thoughts and contributions! Looks great. A couple responses:

- > I changed the name Parent which was always a big cryptic to me to similar_FlatSkyMap which mirrors Julia similar, in that it creates and new and empty object similar to the thing you pass in.

"Parent" was chosen to mirror `spt3g_software`'s convention of using `parent` to describe (often empty) map stubs that are used to hold metadata for `MapSpectrum2D`s. I'm happy to keep this change, it will make things less cryptic for Julia users but maybe a little more cryptic for 3G-python users?

- > It seemed like in some places conversion factors were getting applied in one direction but not the other, such that a field wasn't surviving a round trip to Python, not sure if that was intentional. I made it consistent and added tests to reflect that too.

Awesome, thanks so much for writing tests--I hadn't done any kind of numerical checking yet, so I'm not surprised there were some inconsistencies. But looking at your changes, I can't actually find where you fixed this--out of curiosity, can you point me to the lines in question?

- > Currently we only autoconvert the first two, should we also do the latter?

Ah yes good point, `MapSpectraTEB`s convert to `FlatIEBFourier`s, but `FlatIEBFourier`s convert to `dicts`. But do we really want `FlatIEBFourier`s to be "special" compared to other `FlatS02`s (that all convert to dicts)? It seems weird for `jl("f::FlatIEBFourier")` to be a different type of object with different arithmetic properties than `jl("f::FlatIQUFourier")`. You can always do `jl("MapSpectraTEB(f)")` when you want that. I guess it is a question of preserving back-and-forth autoconversion _consistency_ vs preserving autoconversion _symmetry_? If that makes sense

cailmdaley commented:
> I guess the main thing left is maybe units. I like your suggestion above, but maybe we should be more generic and call it `py"unitless(f)"` and `jl("unitless(f)")'`, where by default (as we are I think) we assume a unitful FFT and convert between the two conventions, and `unitless` overrides that default.
> 
> However, this makes me think, 1) a transfer function will have no units, 2) the FFT of a field will have some units, and 3) a covariance like `diag(Cf)` should have those units squared, so it seems like we should technically need 3 different cases? I'm slightly confused why it doesn't appear we've needed that thus far. Maybe we only ever used 1 and 3, and otherwise were doing conversion in map space.

`unitless` is a much better name! And good point about covariances. Maybe we could have `unitless` dispatch on`LinOp` (or whatever the correct umbrella type for covariances is?) vs. `FlatFourier`. I am actually not sure if we ever converted covariance objects to python in Ultradeep, so it may be that we've only done 1) and 2)?

marius311 committed:
autoconvert FlatIEBFourier, cleanup error messages

marius311 commented:
>Ah yes good point, MapSpectraTEBs convert to FlatIEBFouriers, but FlatIEBFouriers convert to dicts. But do we really want FlatIEBFouriers to be "special" compared to other FlatS02s (that all convert to dicts)? It seems weird for jl("f::FlatIEBFourier") to be a different type of object with different arithmetic properties than jl("f::FlatIQUFourier").

Slightly weird maybe, but I view that more as a limitation on the 3G side since it doesn't have objects for all the different types. As far as this interface goes, I'd say we just convert to the best thing possible, so I added this autoconversion as well.

>Is there a reason to restrict a user from going from any FlatS02 to a python MapSpectraTEB in one step? Calling e.g. MapSpectraTEB(f::FlatIQUMap)(which worked before) is convenient

The problem is that if we want these kinds of calls to also auto-convert basis, there's like an explosion of cases we also logically want and need to implement, many of which don't currently work, e.g. `FlatSkyMap(f::FlatFourier)`, `FlatMap(f::py"MapSpectrum2D")`, etc...  So I think its cleaner to just have these explicit calls only work on the correct type, and if you want your one step conversion, given that I added the auto-conversion thing above, you can do `jl("IEBFourier(f)")` and you will get a `MapSpectraTEB` independent of whatever basis `f` may be in. 

>>It seemed like in some places conversion factors were getting applied in one direction but not the other, such that a field wasn't surviving a round trip to Python, not sure if that was intentional. I made it consistent and added tests to reflect that too.
>
> Awesome, thanks so much for writing tests--I hadn't done any kind of numerical checking yet, so I'm not surprised there were some inconsistencies. But looking at your changes, I can't actually find where you fixed this--out of curiosity, can you point me to the lines in question?

Sorry, looking back it wasnt the scale factor, I think the only thing was that `FlatMap(f::PyObject)` wasn't flipping y.

marius311 committed:
workaround until PyCall PR is merged

marius311 committed:
make valid Python package

marius311 committed:
add readme

marius311 commented:
Ok, ready from my end except for the unitless stuff. Let me know if you have any other comments / responses. 

Also, once merged, lets rename the repo `CMBLensingSPT3GInterface` (the readme reflects this btw).

marius311 committed:
add unitless

marius311 committed:
bidirectional units handled correctly & have tests

marius311 committed:
add copy(order='C')

marius311 committed:
fix Q/U convention difference

marius311 committed:
remove flipping

marius311 commented:
Ok, after further thought, mirroring y in the auto conversion is hopeless for the following reason:

Mirroring a map should mirror E and mirror _and negate_ B (this is basically by definition what makes B be B). But mirroring the diagonal of a covariance operator like `Cf = <f f^T>` should just mirror both E and B without the negation (since two negatives in the B component cancel). This would technically be fine to keep track of in CMBLensing since it has separate datastructures for fields and for fields which are the diagonal of operators, but since spt3g_software doesn't have any way to track whether a field is just a field or whether its the diagonal of an operator, there's no way to know and to negate or not B in the auto-conversion. Hence, hopeless. 

So in the end you're right, easiest thing is just forget about the flipping and plot with `origin="lower"`. This is ready to merge from my end whenever you're ready. I've checked and this works with the QE notebook (https://github.com/SouthPoleTelescope/ultradeep100d/commit/80472b2d059b4843583ea201b200adfb3fb4c486), and the simplifications to the code there are nice, e.g.:

![image](https://user-images.githubusercontent.com/1510968/88649888-18930880-d07d-11ea-8768-b387024b4c94.png)

cailmdaley committed:
removed py tests, added note about Julia environments to README

cailmdaley committed:
actually edit README this time

cailmdaley commented:
Great, I like all these changes, and thanks for taking a crack at the flipping one last time. I'll merge and rename the repo now!
